DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1CompareVertexSoA.cc

```diff
@@ -0,0 +1,156 @@
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  const edm::EDGetTokenT<ZVertexHeterogeneous> tokenSoAVertexCPU_;
+  const edm::EDGetTokenT<ZVertexHeterogeneous> tokenSoAVertexGPU_;
+  const edm::EDGetTokenT<reco::BeamSpot> tokenBeamSpot_;
+  const std::string topFolderName_;
+  const float dzCut_;
+  MonitorElement* hnVertex;
+  MonitorElement* hx;
+  MonitorElement* hy;
+  MonitorElement* hz;
+  MonitorElement* hchi2;
+  MonitorElement* hchi2oNdof;
+  MonitorElement* hptv2;
+  MonitorElement* hntrks;
+};
+
+//
+// constructors
+//
+
+SiPixelPhase1CompareVertexSoA::SiPixelPhase1CompareVertexSoA(const edm::ParameterSet& iConfig)
+    : tokenSoAVertexCPU_(consumes<ZVertexHeterogeneous>(iConfig.getParameter<edm::InputTag>("pixelVertexSrcCPU"))),
+      tokenSoAVertexGPU_(consumes<ZVertexHeterogeneous>(iConfig.getParameter<edm::InputTag>("pixelVertexSrcGPU"))),
+      tokenBeamSpot_(consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamSpotSrc"))),
+      topFolderName_(iConfig.getParameter<std::string>("TopFolderName")),
+      dzCut_(iConfig.getParameter<double>("dzCut")) {}
+
+//
+// -- Analyze
+//
+void SiPixelPhase1CompareVertexSoA::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& vsoaHandleCPU = iEvent.getHandle(tokenSoAVertexCPU_);
+  const auto& vsoaHandleGPU = iEvent.getHandle(tokenSoAVertexGPU_);
+  if (!vsoaHandleCPU.isValid() || !vsoaHandleGPU.isValid()) {
+    edm::LogWarning("SiPixelPhase1MonitorTrackSoA")
+        << "Either vertex SoA found GPU or CPU not found. Comparison not run!" << std::endl;
+    return;
+  }
+
+  auto const& vsoaCPU = *((vsoaHandleCPU.product())->get());
+  int nVerticesCPU = vsoaCPU.nvFinal;
+  auto const& vsoaGPU = *((vsoaHandleGPU.product())->get());
+  int nVerticesGPU = vsoaGPU.nvFinal;
+
+  auto bsHandle = iEvent.getHandle(tokenBeamSpot_);
+  float x0 = 0., y0 = 0., z0 = 0., dxdz = 0., dydz = 0.;
+  if (!bsHandle.isValid()) {
+    edm::LogWarning("PixelVertexProducer") << "No beamspot found. returning vertexes with (0,0,Z) ";
+  } else {
+    const reco::BeamSpot& bs = *bsHandle;
+    x0 = bs.x0();
+    y0 = bs.y0();
+    z0 = bs.z0();
+    dxdz = bs.dxdz();
+    dydz = bs.dydz();
+  }
+
+  for (int ivc = 0; ivc < nVerticesCPU; ivc++) {
+    auto sic = vsoaCPU.sortInd[ivc];
+    auto zc = vsoaCPU.zv[sic];
+    auto xc = x0 + dxdz * zc;
+    auto yc = y0 + dydz * zc;
+    zc += z0;
+
+    auto ndofCPU = vsoaCPU.ndof[sic];
+    auto chi2CPU = vsoaCPU.chi2[sic];
+    for (int ivg = 0; ivg < nVerticesGPU; ivg++) {
+      auto sig = vsoaGPU.sortInd[ivg];
+      auto zg = vsoaGPU.zv[sig];
+      auto xg = x0 + dxdz * zg;
+      auto yg = y0 + dydz * zg;
+      zg += z0;
+      auto ndofGPU = vsoaGPU.ndof[sig];
+      auto chi2GPU = vsoaGPU.chi2[sig];
+
+      //insert some matching condition
+      if (std::abs(zc - zg) > dzCut_)
+        continue;
```

Comment:
If I understand correctly the logic, currently _all_ combinations of gpu and cpu vertices within `dzCut_` are considered -- not only the closest pair.
While for the tracks only the closest gpu and cpu tracks are considered.
Would it make sense to change the logic here as well ?

Reply:
yes , I think it makes sense.
Done at [76b4e67](https://github.com/cms-sw/cmssw/pull/37619/commits/76b4e679e9fa98f05f5fc7095d00eb4b03926ef5)